### PR TITLE
Repository encloses StateStore, PinStore

### PR DESCRIPTION
**Pull request by ukstv**

### CODE REVIEW

File: packages/core/src/pubsub/message-bus.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Multiplexing IPFS Pubsub.
+ */
+export class MessageBus extends Observable<PubsubMessage> implements SubscriptionLike {
+  private readonly pubsubSubscription: Subscription;
+  private readonly feed$: Subject<PubsubMessage> = new Subject<PubsubMessage>();
+
+  constructor(
+    readonly pubsub: Observable<PubsubMessage> & { next: (m: PubsubMessage) => Subscription },
```
Comment (stbrody) [inline packages/core/src/pubsub/message-bus.ts:15]: woah, what does this syntax do?

File: packages/core/src/state-management/repository.ts [hunk 2]
```diff
@@ -22,30 +26,65 @@ export class Repository {
...
+
+  async has(docId: DocID): Promise<boolean> {
+    const fromMemory = await this.fromMemory(docId);
+    if (fromMemory) return true;
+    const fromState = await this.#pinStore.stateStore.load(docId);
+    return Boolean(fromState);
   }
 
   /**
    * Return a document, either from cache or re-constructed from state store.
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:79]:
```suggestion
   * Return a document, either from cache or re-constructed from state store, but will not load from the network.
```

File: packages/core/src/state-management/repository.ts [hunk 3]
```diff
@@ -22,30 +26,65 @@ export class Repository {
...
+  }
+
+  async fromNetwork(docId: DocID, opts: DocOpts = {}): Promise<Document> {
+    const document = await this.#networkLoad.load(docId);
+    await this.add(document);
+    await document._syncDocumentToCurrent({ ...DEFAULT_LOAD_DOCOPTS, ...opts });
+    return document;
+  }
+
+  async load(docId: DocID, opts: DocOpts = {}): Promise<Document> {
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:61]:
```suggestion
  /**
  * Returns a document from wherever we can get information about it.
  * Starts by checking if the document state is present in the in-memory cache, if not then then checks the state store, and finally loads the document from pubsub.
  */
  async load(docId: DocID, opts: DocOpts = {}): Promise<Document> {
```

File: packages/core/src/state-management/repository.ts [hunk 4]
```diff
@@ -22,30 +26,65 @@ export class Repository {
...
+    return this.sync.run(docId.toString(), async () => {
+      const fromMemory = await this.fromMemory(docId);
+      if (fromMemory) return fromMemory;
+      const fromStateStore = await this.fromStateStore(docId);
+      if (fromStateStore) return fromStateStore;
+      return this.fromNetwork(docId, opts);
+    });
+  }
+
+  async has(docId: DocID): Promise<boolean> {
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:71]:
```suggestion
  /**
  * Checks if we can get the document state without having to load it via pubsub (i.e. we have the document state in our in-memory cache or in the state store)
  */
  async has(docId: DocID): Promise<boolean> {
```

File: packages/core/src/state-management/repository.ts [hunk 5]
```diff
@@ -77,9 +124,25 @@ export class Repository {
     await this.#map.delete(docId.toString());
   }
 
+  async list(docId?: DocID): Promise<string[]> {
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:127]: Can we call this `listPinned` to be clearer?

File: packages/core/src/ceramic.ts [hunk 6]
```diff
@@ -208,11 +206,11 @@ class Ceramic implements CeramicApi {
 
     this._doctypeHandlers = new HandlersMap(this._logger)
 
-    this._repository = modules.repository
-    this._repository.setStateStore(this.pinStore.stateStore)
-    const documentFactory = new DocumentFactory(this.dispatcher, this.pinStore, this.context, this._validateDocs, this._doctypeHandlers)
-    this._repository.setDocumentFactory(documentFactory)
-    this.loadingQueue = new LoadingQueue(this._repository, this.dispatcher, this._doctypeHandlers, this.context, this.pinStore, this._logger, documentFactory)
+    const documentFactory = new DocumentFactory(this.dispatcher, pinStore, this.context, this._validateDocs, this._doctypeHandlers)
```
Comment (stbrody) [inline packages/core/src/ceramic.ts:209]: Ideally we'd eventually be in the world where literally nothing besides the `Repository` knows about the `pinStore`.  Looks like `DocumentFactory` and `Document` both still access `pinStore` directly, but it seems like both of those are going away or getting majorly re-worked before this project is done, so maybe their use of `pinStore` will go away naturally in a future commit?

**Review by stbrody — APPROVED** [review]
One question and a few small suggestions, but overall looks great!

File: packages/core/src/dispatcher.ts [hunk 7]
```diff
@@ -197,6 +178,7 @@ export class Dispatcher {
       const document = await this.repository.get(expectedDocID)
       if (document) {
         await document._update(newTip)
+        this._outstandingQueryIds.delete(queryId)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/dispatcher.ts:181]: Uhums.

File: packages/core/src/dispatcher.ts [hunk 8]
```diff
@@ -24,30 +25,16 @@ const IPFS_RESUBSCRIBE_INTERVAL_DELAY = 1000 * 15 // 15 sec
...
-  register (document: Document): void {
-    // Build a QUERY message to send to the pub/sub topic to request the latest tip for this document
-    const message = buildQueryMessage(document.id)
-
-    // Store the query id so we'll process the corresponding RESPONSE message when it comes in
-    this._outstandingQueryIds[message.id] = document.id
-    this.publish(message)
+    const pubsub = new Pubsub(_ipfs, topic, IPFS_RESUBSCRIBE_INTERVAL_DELAY, _pubsubLogger, _logger)
+    this._outstandingQueryIds = new Map();
+    this.messageBus = new MessageBus(pubsub, this._outstandingQueryIds)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/dispatcher.ts:36]: It can.

File: packages/core/src/document.ts [hunk 9]
```diff
@@ -173,8 +160,11 @@ export class Document implements DocStateHolder {
     if (publish) {
       this._publishTip()
     }
+    const tip$ = this.dispatcher.messageBus.queryNetwork(this.id)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/document.ts:163]: Okay.

File: packages/core/src/pubsub/message-bus.ts (hunk 1 above)
Comment (ukstv) [reply packages/core/src/pubsub/message-bus.ts:15]: It states `pubsub` is an Observable _and_ has `next` method. I intend to give a `SubjectLike` name to this interface at the next iteration.

File: packages/core/src/state-management/repository.ts (hunk 5 above)
Comment (ukstv) [reply packages/core/src/state-management/repository.ts]: k

File: packages/core/src/ceramic.ts (hunk 6 above)
Comment (ukstv) [reply packages/core/src/ceramic.ts:209]: Almost every state change has to be stored as `Document#_updateStateIfPinned`, so no way _for now_ this goes away. One could get rid of it at a later stage, after we have a running state.

**Pull request merged**